File: scripts/validate_data.py
#!/usr/bin/env python3
"""
Data validation script for db.json

Performs comprehensive validation checks on the database including:
- JSON validity
- Capitalization patterns for band names, genres, and roles
- Duplicate detection using Levenshtein distance
"""

import json
import sys
from pathlib import Path
from typing import Dict, List, Set, Tuple


# ANSI color codes for terminal output
class Colors:
    """ANSI color codes for colorized terminal output"""

    HEADER = "\033[95m"
    BLUE = "\033[94m"
    CYAN = "\033[96m"
    GREEN = "\033[92m"
    YELLOW = "\033[93m"
    RED = "\033[91m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"
    END = "\033[0m"


def colorize(text: str, color: str) -> str:
    """Colorize text with ANSI codes"""
    return f"{color}{text}{Colors.END}"


def print_header(text: str):
    """Print a colorized header"""
    print(f"\n{colorize('═' * 80, Colors.BOLD)}")
    print(colorize(f"  {text}", Colors.BOLD + Colors.CYAN))
    print(f"{colorize('═' * 80, Colors.BOLD)}\n")


def print_success(text: str):
    """Print success message"""
    print(f"{colorize('✓', Colors.GREEN)} {text}")


def print_warning(text: str):
    """Print warning message"""
    print(f"{colorize('⚠', Colors.YELLOW)} {colorize(text, Colors.YELLOW)}")


def print_error(text: str):
    """Print error message"""
    print(f"{colorize('✗', Colors.RED)} {colorize(text, Colors.RED)}")


def print_info(text: str):
    """Print info message"""
    print(f"{colorize('ℹ', Colors.BLUE)} {text}")


def levenshtein_distance(s1: str, s2: str) -> int:
    """
    Calculate the Levenshtein distance between two strings.

    Args:
        s1: First string
        s2: Second string

    Returns:
        The Levenshtein distance (number of single-character edits needed)
    """
    if len(s1) < len(s2):
        return levenshtein_distance(s2, s1)

    if len(s2) == 0:
        return len(s1)

    previous_row = range(len(s2) + 1)
    for i, c1 in enumerate(s1):
        current_row = [i + 1]
        for j, c2 in enumerate(s2):
            # Cost of insertions, deletions, or substitutions
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1
            substitutions = previous_row[j] + (c1 != c2)
            current_row.append(min(insertions, deletions, substitutions))
        previous_row = current_row

    return previous_row[-1]


def is_capitalized_per_word(text: str, allow_exceptions: bool = True) -> bool:
    """
    Check if text follows capitalized per word pattern.

    Args:
        text: The text to check
        allow_exceptions: Allow common exceptions like 'of', 'the', 'and', 'in', 'a', etc.

    Returns:
        True if the text follows the pattern, False otherwise
    """
    # Common exceptions that can be lowercase in titles
    exceptions = {
        "of",
        "the",
        "and",
        "in",
        "a",
        "an",
        "or",
        "but",
        "for",
        "at",
        "by",
        "to",
        "from",
        "with",
    }

    words = text.split()
    if not words:
        return False

    for i, word in enumerate(words):
        # Skip empty words
        if not word:
            continue

        # Skip words that are all special characters or numbers
        if word.isdigit():
            continue

        # Check if word is an exception (but not the first word)
        if allow_exceptions and i > 0 and word.lower() in exceptions:
            if word.islower():
                continue

        # For non-exception words, check if first letter is capitalized
        first_letter_idx = next((j for j, c in enumerate(word) if c.isalpha()), None)
        if first_letter_idx is not None:
            if not word[first_letter_idx].isupper():
                return False

    return True


def is_role_capitalized(role: str) -> bool:
    """
    Check if band member role has first word capitalized.

    Args:
        role: The role text to check

    Returns:
        True if properly capitalized, False otherwise
    """
    if not role:
        return False

    # Get first word
    first_word = role.split()[0] if role.split() else role

    # Find first letter
    first_letter_idx = next((i for i, c in enumerate(first_word) if c.isalpha()), None)
    if first_letter_idx is None:
        return True  # No letters, skip validation

    return first_word[first_letter_idx].isupper()


def validate_json_structure(file_path: Path) -> Tuple[bool, Dict]:
    """
    Validate that the file is valid JSON and load it.

    Args:
        file_path: Path to the JSON file

    Returns:
        Tuple of (success, data or None)
    """
    print_header("JSON STRUCTURE VALIDATION")

    try:
        with open(file_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        print_success(f"Valid JSON structure")
        return True, data
    except json.JSONDecodeError as e:
        print_error(f"Invalid JSON: {e}")
        return False, {}
    except FileNotFoundError:
        print_error(f"File not found: {file_path}")
        return False, {}
    except Exception as e:
        print_error(f"Error reading file: {e}")
        return False, {}


def validate_band_names(data: Dict) -> Tuple[int, int]:
    """
    Validate band names in both festivals and bands sections.

    Args:
        data: The loaded JSON data

    Returns:
        Tuple of (errors, warnings)
    """
    print_header("BAND NAME CAPITALIZATION")

    errors = 0
    warnings = 0

    # Check band names in festivals
    print_info("Checking band names in festivals...")
    for festival in data.get("festivals", []):
        festival_name = festival.get("name", "Unknown")
        for band_name in festival.get("bands", []):
            if not is_capitalized_per_word(band_name, allow_exceptions=True):
                print_error(
                    f"  Festival '{festival_name}': Band name '{band_name}' not properly capitalized"
                )
                errors += 1

    if errors == 0:
        print_success(
            f"All {sum(len(f.get('bands', [])) for f in data.get('festivals', []))} band names in festivals are properly capitalized"
        )

    # Check band names in bands section
    print_info("Checking band names in bands section...")
    bands_errors = 0
    for band in data.get("bands", []):
        band_name = band.get("name", "")
        if not is_capitalized_per_word(band_name, allow_exceptions=True):
            print_error(f"  Band '{band_name}' not properly capitalized")
            bands_errors += 1
            errors += 1

    if bands_errors == 0:
        print_success(
            f"All {len(data.get('bands', []))} band names are properly capitalized"
        )

    return errors, warnings


def validate_genres(data: Dict) -> Tuple[int, int]:
    """
    Validate music genre capitalization.

    Args:
        data: The loaded JSON data

    Returns:
        Tuple of (errors, warnings)
    """
    print_header("MUSIC GENRE CAPITALIZATION")

    errors = 0
    warnings = 0

    for band in data.get("bands", []):
        band_name = band.get("name", "Unknown")
        for genre in band.get("genres", []):
            if not is_capitalized_per_word(genre, allow_exceptions=False):
                print_error(
                    f"  Band '{band_name}': Genre '{genre}' not properly capitalized"
                )
                errors += 1

    if errors == 0:
        total_genres = sum(
            len(band.get("genres", [])) for band in data.get("bands", [])
        )
        print_success(f"All {total_genres} genre entries are properly capitalized")

    return errors, warnings


def validate_member_roles(data: Dict) -> Tuple[int, int]:
    """
    Validate band member role capitalization.

    Args:
        data: The loaded JSON data

    Returns:
        Tuple of (errors, warnings)
    """
    print_header("BAND MEMBER ROLE CAPITALIZATION")

    errors = 0
    warnings = 0

    for band in data.get("bands", []):
        band_name = band.get("name", "Unknown")
        for member in band.get("members", []):
            member_name = member.get("name", "Unknown")
            role = member.get("role", "")

            if not is_role_capitalized(role):
                print_error(
                    f"  Band '{band_name}', Member '{member_name}': Role '{role}' first word not capitalized"
                )
                errors += 1

    if errors == 0:
        total_members = sum(
            len(band.get("members", [])) for band in data.get("bands", [])
        )
        print_success(f"All {total_members} member roles are properly capitalized")

    return errors, warnings


def detect_duplicates(data: Dict, threshold: int = 2) -> Tuple[int, int]:
    """
    Detect potential duplicate band names using Levenshtein distance.

    Args:
        data: The loaded JSON data
        threshold: Maximum Levenshtein distance to consider as potential duplicate

    Returns:
        Tuple of (errors, warnings)
    """
    print_header("DUPLICATE DETECTION (Levenshtein Distance)")

    errors = 0
    warnings = 0

    # Collect all band names from both sections
    all_bands: List[Tuple[str, str]] = []  # (name, source)

    # From festivals
    for festival in data.get("festivals", []):
        festival_name = festival.get("name", "Unknown")
        for band_name in festival.get("bands", []):
            all_bands.append((band_name, f"Festival: {festival_name}"))

    # From bands section
    for band in data.get("bands", []):
        band_name = band.get("name", "")
        all_bands.append((band_name, "Bands section"))

    # Check for duplicates
    print_info(
        f"Checking {len(all_bands)} band entries for potential duplicates (threshold: {threshold})..."
    )

    checked_pairs: Set[Tuple[str, str]] = set()
    duplicates_found = 0

    for i, (name1, source1) in enumerate(all_bands):
        for j, (name2, source2) in enumerate(all_bands):
            if i >= j:
                continue

            # Normalize for comparison
            normalized_pair = tuple(sorted([name1.lower(), name2.lower()]))
            if normalized_pair in checked_pairs:
                continue

            checked_pairs.add(normalized_pair)

            # Skip if exactly the same (legitimate duplicates across festivals)
            if name1 == name2:
                continue

            distance = levenshtein_distance(name1.lower(), name2.lower())

            if distance <= threshold:
                print_warning(
                    f"  Potential duplicate (distance={distance}): '{name1}' ({source1}) ↔ '{name2}' ({source2})"
                )
                warnings += 1
                duplicates_found += 1

    if duplicates_found == 0:
        print_success("No potential duplicates detected")
    else:
        print_info(f"Found {duplicates_found} potential duplicate(s)")

    return errors, warnings


def main():
    """Main validation function"""
    print(
        colorize("\n🎸 Metal Festivals Database Validator", Colors.BOLD + Colors.CYAN)
    )
    print(colorize("=" * 80, Colors.BOLD))

    # Determine file path
    db_path = Path(__file__).parent.parent / "db.json"

    # Validate JSON structure
    success, data = validate_json_structure(db_path)
    if not success:
        print_error("\n❌ Validation failed: Invalid JSON structure")
        sys.exit(1)

    # Print data summary
    festivals_count = len(data.get("festivals", []))
    bands_count = len(data.get("bands", []))
    print_info(f"Loaded {festivals_count} festivals and {bands_count} bands")

    # Run all validation checks
    total_errors = 0
    total_warnings = 0

    errors, warnings = validate_band_names(data)
    total_errors += errors
    total_warnings += warnings

    errors, warnings = validate_genres(data)
    total_errors += errors
    total_warnings += warnings

    errors, warnings = validate_member_roles(data)
    total_errors += errors
    total_warnings += warnings

    errors, warnings = detect_duplicates(data, threshold=2)
    total_errors += errors
    total_warnings += warnings

    # Print summary
    print_header("VALIDATION SUMMARY")

    if total_errors == 0 and total_warnings == 0:
        print(
            colorize(
                "✅ All validations passed! No issues found.",
                Colors.BOLD + Colors.GREEN,
            )
        )
        sys.exit(0)
    else:
        if total_errors > 0:
            print(
                colorize(f"❌ Found {total_errors} error(s)", Colors.BOLD + Colors.RED)
            )
        if total_warnings > 0:
            print(
                colorize(
                    f"⚠️  Found {total_warnings} warning(s)", Colors.BOLD + Colors.YELLOW
                )
            )

        if total_errors > 0:
            sys.exit(1)
        else:
            sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/validate_data.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"unicode"
)

// ANSI color codes for terminal output
const (
	ColorHeader    = "\033[95m"
	ColorBlue      = "\033[94m"
	ColorCyan      = "\033[96m"
	ColorGreen     = "\033[92m"
	ColorYellow    = "\033[93m"
	ColorRed       = "\033[91m"
	ColorBold      = "\033[1m"
	ColorUnderline = "\033[4m"
	ColorEnd       = "\033[0m"
)

// Database structures
type Database struct {
	Festivals []Festival `json:"festivals"`
	Bands     []Band     `json:"bands"`
}

type Festival struct {
	Name  string   `json:"name"`
	Bands []string `json:"bands"`
}

type Band struct {
	Name    string   `json:"name"`
	Genres  []string `json:"genres"`
	Members []Member `json:"members"`
}

type Member struct {
	Name string `json:"name"`
	Role string `json:"role"`
}

// ValidationResult tracks errors and warnings
type ValidationResult struct {
	Errors   int
	Warnings int
}

// Helper functions for colorized output
func colorize(text, color string) string {
	return fmt.Sprintf("%s%s%s", color, text, ColorEnd)
}

func printHeader(text string) {
	fmt.Printf("\n%s\n", colorize(strings.Repeat("═", 80), ColorBold))
	fmt.Printf("%s\n", colorize(fmt.Sprintf("  %s", text), ColorBold+ColorCyan))
	fmt.Printf("%s\n\n", colorize(strings.Repeat("═", 80), ColorBold))
}

func printSuccess(text string) {
	fmt.Printf("%s %s\n", colorize("✓", ColorGreen), text)
}

func printWarning(text string) {
	fmt.Printf("%s %s\n", colorize("⚠", ColorYellow), colorize(text, ColorYellow))
}

func printError(text string) {
	fmt.Printf("%s %s\n", colorize("✗", ColorRed), colorize(text, ColorRed))
}

func printInfo(text string) {
	fmt.Printf("%s %s\n", colorize("ℹ", ColorBlue), text)
}

// levenshteinDistance calculates the Levenshtein distance between two strings
func levenshteinDistance(s1, s2 string) int {
	r1 := []rune(s1)
	r2 := []rune(s2)

	if len(r1) < len(r2) {
		return levenshteinDistance(s2, s1)
	}

	if len(r2) == 0 {
		return len(r1)
	}

	previousRow := make([]int, len(r2)+1)
	for i := range previousRow {
		previousRow[i] = i
	}

	for i, c1 := range r1 {
		currentRow := []int{i + 1}
		for j, c2 := range r2 {
			insertions := previousRow[j+1] + 1
			deletions := currentRow[j] + 1
			substitutions := previousRow[j]
			if c1 != c2 {
				substitutions++
			}
			currentRow = append(currentRow, min(insertions, deletions, substitutions))
		}
		previousRow = currentRow
	}

	return previousRow[len(r2)]
}

func min(a, b, c int) int {
	if a < b {
		if a < c {
			return a
		}
		return c
	}
	if b < c {
		return b
	}
	return c
}

// isCapitalizedPerWord checks if text follows capitalized per word pattern
func isCapitalizedPerWord(text string, allowExceptions bool) bool {
	exceptions := map[string]bool{
		"of": true, "the": true, "and": true, "in": true, "a": true,
		"an": true, "or": true, "but": true, "for": true, "at": true,
		"by": true, "to": true, "from": true, "with": true,
	}

	words := strings.Fields(text)
	if len(words) == 0 {
		return false
	}

	for i, word := range words {
		if word == "" {
			continue
		}

		// Skip words that are all digits
		allDigits := true
		for _, c := range word {
			if !unicode.IsDigit(c) {
				allDigits = false
				break
			}
		}
		if allDigits {
			continue
		}

		// Check if word is an exception (but not the first word)
		if allowExceptions && i > 0 && exceptions[strings.ToLower(word)] {
			if strings.ToLower(word) == word {
				continue
			}
		}

		// For non-exception words, check if first letter is capitalized
		firstLetterIdx := -1
		for j, c := range word {
			if unicode.IsLetter(c) {
				firstLetterIdx = j
				break
			}
		}

		if firstLetterIdx != -1 {
			firstLetter := []rune(word)[firstLetterIdx]
			if !unicode.IsUpper(firstLetter) {
				return false
			}
		}
	}

	return true
}

// isRoleCapitalized checks if band member role has first word capitalized
func isRoleCapitalized(role string) bool {
	if role == "" {
		return false
	}

	words := strings.Fields(role)
	if len(words) == 0 {
		return false
	}

	firstWord := words[0]

	// Find first letter
	firstLetterIdx := -1
	for i, c := range firstWord {
		if unicode.IsLetter(c) {
			firstLetterIdx = i
			break
		}
	}

	if firstLetterIdx == -1 {
		return true // No letters, skip validation
	}

	firstLetter := []rune(firstWord)[firstLetterIdx]
	return unicode.IsUpper(firstLetter)
}

// validateJSONStructure validates that the file is valid JSON and loads it
func validateJSONStructure(filePath string) (bool, *Database) {
	printHeader("JSON STRUCTURE VALIDATION")

	file, err := os.ReadFile(filePath)
	if err != nil {
		printError(fmt.Sprintf("Error reading file: %v", err))
		return false, nil
	}

	var data Database
	if err := json.Unmarshal(file, &data); err != nil {
		printError(fmt.Sprintf("Invalid JSON: %v", err))
		return false, nil
	}

	printSuccess("Valid JSON structure")
	return true, &data
}

// validateBandNames validates band names in both festivals and bands sections
func validateBandNames(data *Database) ValidationResult {
	printHeader("BAND NAME CAPITALIZATION")

	result := ValidationResult{}

	// Check band names in festivals
	printInfo("Checking band names in festivals...")
	totalFestivalBands := 0
	for _, festival := range data.Festivals {
		for _, bandName := range festival.Bands {
			totalFestivalBands++
			if !isCapitalizedPerWord(bandName, true) {
				printError(fmt.Sprintf("  Festival '%s': Band name '%s' not properly capitalized", festival.Name, bandName))
				result.Errors++
			}
		}
	}

	if result.Errors == 0 {
		printSuccess(fmt.Sprintf("All %d band names in festivals are properly capitalized", totalFestivalBands))
	}

	// Check band names in bands section
	printInfo("Checking band names in bands section...")
	bandsErrors := 0
	for _, band := range data.Bands {
		if !isCapitalizedPerWord(band.Name, true) {
			printError(fmt.Sprintf("  Band '%s' not properly capitalized", band.Name))
			bandsErrors++
			result.Errors++
		}
	}

	if bandsErrors == 0 {
		printSuccess(fmt.Sprintf("All %d band names are properly capitalized", len(data.Bands)))
	}

	return result
}

// validateGenres validates music genre capitalization
func validateGenres(data *Database) ValidationResult {
	printHeader("MUSIC GENRE CAPITALIZATION")

	result := ValidationResult{}
	totalGenres := 0

	for _, band := range data.Bands {
		for _, genre := range band.Genres {
			totalGenres++
			if !isCapitalizedPerWord(genre, false) {
				printError(fmt.Sprintf("  Band '%s': Genre '%s' not properly capitalized", band.Name, genre))
				result.Errors++
			}
		}
	}

	if result.Errors == 0 {
		printSuccess(fmt.Sprintf("All %d genre entries are properly capitalized", totalGenres))
	}

	return result
}

// validateMemberRoles validates band member role capitalization
func validateMemberRoles(data *Database) ValidationResult {
	printHeader("BAND MEMBER ROLE CAPITALIZATION")

	result := ValidationResult{}
	totalMembers := 0

	for _, band := range data.Bands {
		for _, member := range band.Members {
			totalMembers++
			if !isRoleCapitalized(member.Role) {
				printError(fmt.Sprintf("  Band '%s', Member '%s': Role '%s' first word not capitalized", band.Name, member.Name, member.Role))
				result.Errors++
			}
		}
	}

	if result.Errors == 0 {
		printSuccess(fmt.Sprintf("All %d member roles are properly capitalized", totalMembers))
	}

	return result
}

// detectDuplicates detects potential duplicate band names using Levenshtein distance
func detectDuplicates(data *Database, threshold int) ValidationResult {
	printHeader("DUPLICATE DETECTION (Levenshtein Distance)")

	result := ValidationResult{}

	// Collect all band names from both sections
	type BandEntry struct {
		Name   string
		Source string
	}
	var allBands []BandEntry

	// From festivals
	for _, festival := range data.Festivals {
		for _, bandName := range festival.Bands {
			allBands = append(allBands, BandEntry{
				Name:   bandName,
				Source: fmt.Sprintf("Festival: %s", festival.Name),
			})
		}
	}

	// From bands section
	for _, band := range data.Bands {
		allBands = append(allBands, BandEntry{
			Name:   band.Name,
			Source: "Bands section",
		})
	}

	printInfo(fmt.Sprintf("Checking %d band entries for potential duplicates (threshold: %d)...", len(allBands), threshold))

	checkedPairs := make(map[string]bool)
	duplicatesFound := 0

	for i := 0; i < len(allBands); i++ {
		for j := i + 1; j < len(allBands); j++ {
			name1 := allBands[i].Name
			name2 := allBands[j].Name
			source1 := allBands[i].Source
			source2 := allBands[j].Source

			// Create normalized pair key
			lower1 := strings.ToLower(name1)
			lower2 := strings.ToLower(name2)
			var pairKey string
			if lower1 < lower2 {
				pairKey = lower1 + "|" + lower2
			} else {
				pairKey = lower2 + "|" + lower1
			}

			if checkedPairs[pairKey] {
				continue
			}
			checkedPairs[pairKey] = true

			// Skip if exactly the same (legitimate duplicates across festivals)
			if name1 == name2 {
				continue
			}

			distance := levenshteinDistance(lower1, lower2)

			if distance <= threshold {
				printWarning(fmt.Sprintf("  Potential duplicate (distance=%d): '%s' (%s) ↔ '%s' (%s)", distance, name1, source1, name2, source2))
				result.Warnings++
				duplicatesFound++
			}
		}
	}

	if duplicatesFound == 0 {
		printSuccess("No potential duplicates detected")
	} else {
		printInfo(fmt.Sprintf("Found %d potential duplicate(s)", duplicatesFound))
	}

	return result
}

func main() {
	fmt.Printf("\n%s\n", colorize("🎸 Metal Festivals Database Validator", ColorBold+ColorCyan))
	fmt.Printf("%s\n", colorize(strings.Repeat("=", 80), ColorBold))

	// Determine file path
	execPath, err := os.Executable()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting executable path: %v\n", err)
		os.Exit(1)
	}
	execDir := filepath.Dir(execPath)
	dbPath := filepath.Join(execDir, "..", "db.json")

	// For go run, use current directory
	if strings.Contains(execPath, "go-build") {
		wd, _ := os.Getwd()
		dbPath = filepath.Join(wd, "db.json")
	}

	// Validate JSON structure
	success, data := validateJSONStructure(dbPath)
	if !success {
		printError("\n❌ Validation failed: Invalid JSON structure")
		os.Exit(1)
	}

	// Print data summary
	printInfo(fmt.Sprintf("Loaded %d festivals and %d bands", len(data.Festivals), len(data.Bands)))

	// Run all validation checks
	totalErrors := 0
	totalWarnings := 0

	result := validateBandNames(data)
	totalErrors += result.Errors
	totalWarnings += result.Warnings

	result = validateGenres(data)
	totalErrors += result.Errors
	totalWarnings += result.Warnings

	result = validateMemberRoles(data)
	totalErrors += result.Errors
	totalWarnings += result.Warnings

	result = detectDuplicates(data, 2)
	totalErrors += result.Errors
	totalWarnings += result.Warnings

	// Print summary
	printHeader("VALIDATION SUMMARY")

	if totalErrors == 0 && totalWarnings == 0 {
		fmt.Printf("%s\n", colorize("✅ All validations passed! No issues found.", ColorBold+ColorGreen))
		os.Exit(0)
	} else {
		if totalErrors > 0 {
			fmt.Printf("%s\n", colorize(fmt.Sprintf("❌ Found %d error(s)", totalErrors), ColorBold+ColorRed))
		}
		if totalWarnings > 0 {
			fmt.Printf("%s\n", colorize(fmt.Sprintf("⚠️  Found %d warning(s)", totalWarnings), ColorBold+ColorYellow))
		}

		if totalErrors > 0 {
			os.Exit(1)
		}
		os.Exit(0)
	}
}
